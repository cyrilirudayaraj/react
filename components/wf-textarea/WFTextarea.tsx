import React from 'react';
import { FormField, Textarea, Input } from '@athena/forge';
import Labels from '../../constants/Labels';
import { FormikProps } from 'formik';
import AppConstants from '../../constants/AppConstants';
import { get } from 'lodash';

interface TextareaCounterProps {
  formik: FormikProps<any>;
  mandatoryFields?: any;
  label: any;
  name: any;
  maxlength: any;
  className?: any;
  labelWidth?: any;
  istextarea?: any;
  disabled?: boolean;
}

export default class WFTextarea extends React.Component<TextareaCounterProps> {
  state = {
    expand: false,
    showMore: false,
  };

  toggleMinHeight = (event: any): void => {
    const targetObject = document.getElementById(event.target.id);
    if (targetObject !== null) {
      targetObject.style.maxHeight =
        AppConstants.UI_CONSTANTS.TEXTAREA.MIN_HEIGHT;
      this.setState({
        expand: false,
      });
    }
  };

  toggleMaxHeight = (event: any): void => {
    const targetObject = document.getElementById(event.target.id);
    if (targetObject !== null) {
      this.setState({
        expand: true,
      });
    }
  };

  checkShowMore = (event: any): void => {
    const { name, value } = event.target;
    const { formik } = this.props;
    formik.setFieldValue(name, value);
    const targetObject = document.getElementById(event.target.id);
    if (targetObject?.style) {
      if (
        targetObject?.style.height >
        AppConstants.UI_CONSTANTS.TEXTAREA.MIN_HEIGHT
      ) {
        this.setState({
          showMore: true,
        });
      } else {
        this.setState({
          showMore: false,
        });
      }
    }
  };

  toggleexpand = (event: any, id: any): void => {
    const buttonelement = document.getElementById(event.target.id);
    const textelement = document.getElementById(id);
    if (buttonelement !== null && textelement !== null) {
      if (buttonelement.innerText === Labels.COMPONENTS.TEXTAREA.SHOW_MORE) {
        textelement.style.height = AppConstants.UI_CONSTANTS.TEXTAREA.HEIGHT;
        textelement.style.maxHeight =
          AppConstants.UI_CONSTANTS.TEXTAREA.MAX_HEIGHT;
        buttonelement.innerText = Labels.COMPONENTS.TEXTAREA.SHOW_LESS;
      } else {
        textelement.style.height = AppConstants.UI_CONSTANTS.TEXTAREA.HEIGHT;
        textelement.style.maxHeight =
          AppConstants.UI_CONSTANTS.TEXTAREA.MIN_HEIGHT;
        buttonelement.innerText = Labels.COMPONENTS.TEXTAREA.SHOW_MORE;
      }
    }
  };

  getFormFieldProps = (fieldName: string, formik: any) => {
    const fieldProps = {
      ...formik.getFieldProps(fieldName),
      required: this.props?.mandatoryFields
        ? this.props.mandatoryFields[fieldName]
        : true,
      autoComplete: 'off',
      error: get(formik.errors, fieldName),
      id: fieldName,
      className: this.props.className,
      labelWidth: this.props.labelWidth ? this.props.labelWidth : 1,
    };
    return fieldProps;
  };

  render(): JSX.Element {
    const fieldLength =
      this.props.formik.getFieldProps(this.props.name)?.value?.length || 0;
    return (
      <div className="textarea-component">
        <FormField
          style={{
            borderColor: fieldLength >= this.props.maxlength ? '#FFAB00' : '',
          }}
          inputAs={this.props.istextarea === true ? Textarea : Input}
          labelText={this.props.label}
          {...this.getFormFieldProps(this.props.name, this.props.formik)}
          onChange={(event: any) => this.checkShowMore(event)}
          onBlur={(event: any) => this.toggleMinHeight(event)}
          onFocus={(event: any) => this.toggleMaxHeight(event)}
          disabled={this.props.disabled}
        />
        <div className="counter">
          {this.state.expand === false &&
          this.props.istextarea === true &&
          this.state.showMore === true ? (
            <a
              id={'togglebtn-' + this.props.name}
              onClick={(event: any) =>
                this.toggleexpand(event, this.props.name)
              }
            >
              {Labels.COMPONENTS.TEXTAREA.SHOW_MORE}
            </a>
          ) : (
            ''
          )}
          {this.state.expand === true ? (
            <div>
              {fieldLength}/{this.props.maxlength}
            </div>
          ) : (
            ''
          )}
        </div>
      </div>
    );
  }
}
