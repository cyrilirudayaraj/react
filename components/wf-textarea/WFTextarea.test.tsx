import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import WFTextarea from './WFTextarea';
import { FormField } from '@athena/forge';

configure({ adapter: new Adapter() });

let wrapper: any;
const formik: any = {
  values: { target: 'Test value' },
  touched: {},
  errors: {},
  setFieldValue: jest.fn(),
  getFieldProps: jest.fn(() => {
    return { value: 'test' };
  }),
};
const props = {
  formik: formik,
  mandatoryFields: {},
  label: 'Example',
  name: 'testExample',
  maxlength: 200,
  istextarea: true,
  disabled: false,
};
beforeEach(() => {
  wrapper = shallow(<WFTextarea {...props} />);
});

describe('test <WFTextarea>', () => {
  it('should render as expected', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
  it('execute events', () => {
    const event: any = {
      target: {
        value: 'Change function',
        name: 'test',
        id: 'testClaimExample',
      },
    };
    wrapper.find(FormField).simulate('change', event);
    wrapper.find(FormField).simulate('focus', event);
    wrapper.find(FormField).simulate('blur', event);
  });
});
