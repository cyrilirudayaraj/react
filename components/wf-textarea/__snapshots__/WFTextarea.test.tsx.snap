// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <WFTextarea> should render as expected 1`] = `
"<div className=\\"textarea-component\\">
  <FormField style={{...}} inputAs={{...}} labelText=\\"Example\\" value=\\"test\\" required={false} autoComplete=\\"off\\" error={[undefined]} id=\\"testExample\\" className={[undefined]} labelWidth={1} onChange={[Function: onChange]} onBlur={[Function: onBlur]} onFocus={[Function: onFocus]} disabled={false} />
  <div className=\\"counter\\" />
</div>"
`;
